src/traces/waterfall/layout_attributes.js

```diff
@@ -0,0 +1,50 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+module.exports = {
+    barmode: {
```

Comment:
I'm not sure we should be reusing the `bar` layout attributes, but then again I'm not sure what _should_ happen on subplots with bar and waterfall traces coexisting. To me, it seems safer to have a different set of layout attributes for waterfall traces: `waterfallmode`, `waterfallgap` and `waterfallgroupgap`, I could be wrong though.

Reply:
After a quick talk with @alexcjohnson ,  I think making waterfall-specific trace layout attributes is the way to go for now. That is, these here should be `waterfallmode`, `waterfallgap` and `waterfallgroupgap`.